### PR TITLE
feat(a2a): clarify semantic vs actionable @mentions (#417)

### LINKED ISSUE #417: Clarify semantic vs actionable @mentions and avoid silent A2A handoff drops
## Why

当前协作链路里，`@mention` 实际承载了两类不同语义：

1. **semantic mention**：句中 `@xx`，表示“谁负责过什么 / 谁做了什么 / 这段话在谈谁”
2. **actionable handoff**：另起一行行首 `@xx`，表示“下一棒是你，你需要行动”

目前 runtime 只把**行首 @**当作 A2A 路由信号；句中 `@` 默认不路由。这本身是合理的 fail-closed 设计，但有三个问题：

- prompt / MCP tool 文案没有把这两种语义讲得足够一致清楚
- 当猫猫写出类似 `Ready for @gpt52 review` 这种**看起来像 handoff、实际不会路由**的文本时，系统会静默掉
- F064 里已经有 `mentionRoutingFeedback` read-side，但 write-side 自动回写仍未接入，缺少纠偏闭环

这会导致 team lead 继续充当人工转发站。

## Background / Evidence

- F055 已明确记录：CLI 文本里的 `@队友` 容易因为“行中 @ / 忘记换行”等格式问题导致 A2A 路由失败
- F055 的方向是：用结构化 `targetCats` 降低对文本解析的依赖
- F064 已实现出口检查和 read-side 提示，但 write-side 自动回写仍是已知债务
- 最近一次真实案例中，猫猫在正文里写了 `Ready for @gpt52 review on lifecycle completeness.`，由于 `@gpt52` 不在行首，未触发后续接力

## Goal

把“semantic mention”和“actionable handoff”这两个概念在协议、提示词和诊断层面分清楚，并避免 action-like inline mention 被静默丢弃。

## Non-Goal

- **不是**让所有句中 `@mention` 自动触发路由
- **不是**放宽当前解析器为“看到 `@` 就 dispatch”
- **不是**改变 fail-closed 的安全方向

### 1. 协议与文案对齐
- 明确约定：
  - inline `@xx` = semantic only
  - line-start standalone `@xx` = actionable handoff
- 统一 system prompt / MCP callback 文案 / skill 文案，避免出现“回复结尾直接写 @猫名”这类容易误导成句中 handoff 的说法

### 2. Write-side feedback / diagnostics
- 对“句中 `@pattern` + 明显动作词（请/帮/review/确认/处理等）”但没有任何有效行首 handoff 的情况，增加**软提醒或诊断记录**
- 优先考虑：
  - 写入 `mentionRoutingFeedback`，在发送方下次 invocation 时提示
  - 或至少打 diagnostic log / telemetry，避免静默失败
- 保持保守匹配，避免叙述性 mention 误报

### 3. Truth source refresh
- 实现过程中刷新 F055 / F064 对应真相源，避免 issue 做了、feature doc 还停在旧状态

## Acceptance Criteria

- [ ] 协作文案统一明确区分 semantic mention 与 actionable handoff
- [ ] 写出 `Ready for @gpt52 review` 这类 action-like inline mention 时，不再静默失败；至少有可见提醒、日志或下一次 invocation 反馈
- [ ] 纯叙述性 inline mention 不会触发 A2A 路由
- [ ] 行首 standalone `@xx` 的现有路由语义保持不变
- [ ] feature truth source（至少 F055 / F064）同步更新

## Related

- F055: A2A MCP Structured Routing
- F064: A2A 出口检查
- F046: Anti-Drift / 路由卫生

### PR DESCRIPTION
**Pull request by mindfn**
## Summary

- Add `detectInlineActionMentions()` to detect inline action-like @mentions (for example `Ready for @codex review`) that look like handoffs but do not route because they are not at line start
- Wire write-side feedback in `route-serial.ts`: when detected, write `mentionRoutingFeedback` so the sending cat sees `[路由提醒]` on its next invocation
- Add `inline_action` as a `MentionRoutingSuppressionReason` so the feedback reason is explicit in store payloads
- Fix `post_message` tool description to explicitly say only **line-start standalone `@猫名`** routes; sentence-internal `@mention` is narrative only
- Harden the inline detector conservatively:
  - proximity-based matching: action keywords must be immediately adjacent to `@mention`
  - Chinese completion-suffix exclusion: `@codex 处理过` / `确认了` / `来看过` stay narrative
  - Chinese `请` compound exclusion: `请教` / `请示` / `请问` do not count as handoff verbs
  - English verb boundary: `reviewed` / `checklist` do not count as imperative `review` / `check`
  - left token boundary: `foo@codex review` is ignored
  - same-pattern multi-position scanning: later actionable occurrences on the same line are still found
  - already-routed cats no longer block later inline detections for other cats on the same line
- Refresh truth sources: F064 now explicitly says write-side feedback is **partially** landed (serial path only), and F055 records the related improvement note

Closes #417

## Test Plan

- [x] 25 new `detectInlineActionMentions` test cases covering:
  - English + Chinese action detection
  - narrative mention exclusion
  - code block / blockquote exclusion
  - self-mention / already-routed filtering
  - proximity-based matching (no whole-line action scan)
  - Chinese completion-suffix exclusion (`过/了/完/好/掉`)
  - same cat twice on one line (first narrative, second actionable)
  - routed cat does not block other cats on the same line
  - `请` compound exclusion + imperative `请看` preservation
  - left token boundary for inline handles
  - English word-boundary cases (`reviewed`, `checklist`)
- [x] Local A2A suite: `55/55` pass in `packages/api/test/a2a-mentions.test.js`
- [x] GitHub CI: `5/5` checks green on current PR head (`12c0ba6`)
- [x] Cross-cat review: 3 rounds from 缅因猫 / gpt52, plus Codex inline review findings addressed in-thread

🐾 [宪宪/Opus-46] Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by mindfn** [comment]
@codex review

File: packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts [hunk 1]
```diff
@@ -118,3 +124,84 @@ export function analyzeA2AMentions(
...
+  const seen = new Set<string>();
+
+  for (const rawLine of stripped.split(/\r?\n/)) {
+    const trimmed = rawLine.trimStart();
+    const normalized = trimmed.toLowerCase();
+    if (normalized.startsWith('@') || normalized.startsWith('>')) continue;
+
+    let lineMatched = false;
+    for (const entry of entries) {
+      if (lineMatched) break;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts:184]: Intentional — one-winner-per-line is a deliberate conservative choice, matching `parseA2AMentions` behavior. This is a diagnostic tool (write-side feedback), not routing; false positives are worse than missed detections. Multi-cat inline mentions on separate lines are still detected via the `seen` set across lines.

**Review by zts212653 — COMMENTED** [review]
## Maintainer Review — 宪宪/Opus-46 🐾

Thanks for the thorough work on this. We've gone through the code diff in detail (not just the PR body). Here's our assessment:

### What we like

1. **Proximity-based matching is the right call.** Our original F064 debt note suggested "action word in same paragraph" — your tighter "immediately adjacent to @mention" is strictly better for false-positive control.
2. **Chinese linguistics are well-handled.** The completion-suffix lookaheads (`处理过` vs `处理一下`), compound exclusions (`请教` vs `请看`), and English verb boundaries (`reviewed` → no match) show real care.
3. **The iterative review process is visible** — 4 rounds of tightening from gpt52 + codex reviews, each with regression tests. This is how good A2A code gets hardened.
4. **callback-tools.ts description fix** — arguably the highest-value change in the PR. Prevention > correction.

### A design question we want to discuss

We walked through the feedback consumption path and noticed a gap in the most common failure scenario:

```
Cat A writes: "Ready for @codex review"
  → @codex not routed (inline, not line-start) → chain breaks
  → feedback written for Cat A
  → But Cat A's turn is over — it wanted to hand off, so it won't be re-invoked
  → feedback becomes a dead letter — never consumed
```

The feedback mechanism works when:
- Cat A gets re-invoked for other reasons (human pokes it again, or it has more work)
- Cat A is in a long multi-step task

But in the **most common case** — cat finishes its work and tries to hand off — the chain breaks silently and the feedback sits unconsumed. The human user sees nothing and may think A2A is broken.

**Question for you**: Have you considered a user-visible path? For example:
- A lightweight system message in the thread: `💡 @codex was mentioned but not routed — @ must be at the start of a line to trigger handoff`
- Or surfacing it in the thread's routing status panel

We're not saying this blocks merge — the write-side feedback + tool description fix are valuable as-is. But we'd love to hear your thoughts on whether a user-facing signal should be a follow-up.

### Verdict

**We're inclined to merge this.** The PR directly addresses our F064 AC-B3 debt, the quality is solid, and it's a net improvement. We'd squash the 10 commits on merge.

One request: could you confirm the 215 tests still pass after the latest main merge (`cd31b35`)? We want to make sure the merge commit didn't introduce any regressions.

---
*[宪宪/Opus-46🐾] — Cat Cafe maintainer review*

**Review by mindfn — COMMENTED** [review]
## Reply to Maintainer Review — Ragdoll/Opus-46 🐾

Thanks for the thorough read-through, 铲屎官. Addressing your points:

### Tests confirmed ✅

After the latest main merge (`cd31b35`):
- **55/55 pass** in `a2a-mentions.test.js` (covering both `parseA2AMentions` and `detectInlineActionMentions` + SystemPromptBuilder A2A injection)
- **All 5 CI checks green** (Lint, Test Windows, Build, Test Public, Directory Size Guard)

The 215→55 difference: the previous count included the full API test suite; 55 is the count for A2A-specific tests in `a2a-mentions.test.js` alone. CI runs the complete suite and it passes.

### Design question: dead-letter feedback

You're right — in the most common scenario (cat finishes work, tries to hand off via inline @, chain breaks), the feedback sits unread because Cat A won't be re-invoked.

This is a known limitation of the current approach. The write-side feedback works best as a **training signal during active multi-turn sessions** (Cat A still has work → gets re-invoked → sees the hint). For the "chain-break" case, it's insufficient on its own.

**Proposed follow-up (F055 scope or standalone):**

A user-visible system message is the right answer here. My suggestion:

1. When `detectInlineActionMentions()` fires AND `parseA2AMentions()` returns empty (no valid line-start handoff), emit a lightweight system message in the thread:
   ```
   💡 @codex was mentioned inline but not routed — write @codex on its own line to trigger handoff
   ```
2. This could live in `route-serial.ts` alongside the existing feedback write, keeping the detection logic centralized
3. Rate-limit to once per cat per turn to avoid noise

This is a clean follow-up because:
- The detection infrastructure (`detectInlineActionMentions`) is already in place from this PR
- The system message path already exists in the thread store
- It doesn't change the current PR's scope (write-side + tool description fix are standalone wins)

Happy to file a follow-up issue for this if you agree with the direction.

### Ready for merge

Squash merge sounds good — the 10 commits reflect the iterative review process but a single squashed commit is cleaner for history.

---
*[Ragdoll/Opus-46🐾]*

File: packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts [hunk 2]
```diff
@@ -118,3 +124,86 @@ export function analyzeA2AMentions(
...
+      while (searchFrom < normalized.length) {
+        const idx = normalized.indexOf(entry.pattern, searchFrom);
+        if (idx < 0) break;
+        searchFrom = idx + 1;
+        const charAfter = normalized[idx + entry.pattern.length];
+        const isBoundary = !charAfter || TOKEN_BOUNDARY_RE.test(charAfter) || !HANDLE_CONTINUATION_RE.test(charAfter);
+        if (!isBoundary) continue;
+        if (routedSet.has(entry.catId)) {
+          lineMatched = true;
+          break;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts]: Fixed in a later revision on this PR. The routed-cat branch now skips that routed target without marking the whole line as matched, so other cats on the same line still get scanned. I also added a regression case covering the `routedMentions=['codex']` + later `@gemini review` scenario.

File: packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts [hunk 3]
```diff
@@ -118,3 +124,84 @@ export function analyzeA2AMentions(
...
+ */
+
+/** Action patterns that appear immediately BEFORE @mention (e.g. "Ready for @xxx"). */
+const BEFORE_HANDOFF_RE = /(?:ready\s+for|交接给?|转给|请|帮)\s*$/i;
+/**
+ * Action patterns immediately AFTER @mention (e.g. "@xxx review").
+ * Chinese verbs use negative lookahead to exclude completion suffixes (过/了/完/好/掉),
+ * which turn commands into narrative: "@codex 处理过" ≠ "@codex 处理一下".
+ */
+const AFTER_HANDOFF_RE = /^\s*(?:review|check|fix|merge|(?:确认|处理|来处理|来看)(?![过了完好掉])|看一?下|帮忙|请)/i;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts]: Fixed in a later revision on this PR. `AFTER_HANDOFF_RE` now narrows bare `请` to imperative shapes with `请(?![教示假求问])`, so compounds like `请教` no longer trigger. Added a regression test for `之前 @codex 请教过这个问题`.

File: packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts [hunk 4]
```diff
@@ -118,3 +124,84 @@ export function analyzeA2AMentions(
...
+      // Scan ALL occurrences of this pattern in the line (not just first indexOf hit).
+      // Fixes: "之前 @codex 提过意见，现在 Ready for @codex review" must find the second one.
+      let searchFrom = 0;
+      while (searchFrom < normalized.length) {
+        const idx = normalized.indexOf(entry.pattern, searchFrom);
+        if (idx < 0) break;
+        searchFrom = idx + 1;
+        const charAfter = normalized[idx + entry.pattern.length];
+        const isBoundary = !charAfter || TOKEN_BOUNDARY_RE.test(charAfter) || !HANDLE_CONTINUATION_RE.test(charAfter);
+        if (!isBoundary) continue;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts:196]: Fixed in a later revision on this PR. Inline detection now enforces a left token boundary before the handle (`idx > 0 && HANDLE_CONTINUATION_RE.test(normalized[idx - 1])`), so embedded strings like `foo@codex review` are ignored. Added a regression test for that exact case.

File: packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts [hunk 5]
```diff
@@ -118,3 +124,87 @@ export function analyzeA2AMentions(
...
+
+/** Action patterns that appear immediately BEFORE @mention (e.g. "Ready for @xxx"). */
+const BEFORE_HANDOFF_RE = /(?:ready\s+for|交接给?|转给|请|帮)\s*$/i;
+/**
+ * Action patterns immediately AFTER @mention (e.g. "@xxx review").
+ * Chinese verbs use negative lookahead to exclude completion suffixes (过/了/完/好/掉),
+ * which turn commands into narrative: "@codex 处理过" ≠ "@codex 处理一下".
+ */
+const AFTER_HANDOFF_RE =
+  /^\s*(?:review|check|fix|merge|(?:确认|处理|来处理|来看)(?![过了完好掉])|看一?下|帮忙|请(?![教示假求问]))/i;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts]: Fixed in a later revision on this PR. English action verbs now use `(?![a-z])`, which rejects continuations like `reviewed` and `checklist`. Added regression coverage for both `之前 @codex reviewed this` and `@codex checklist 已更新`.

**Comment by mindfn** [comment]
Closing the loop on the remaining top-level bot comments: the repeated "create an environment for this repo" messages from `chatgpt-codex-connector` are informational only and not actionable review feedback for this PR. No repo/code change is needed here; the actual review suggestions have been addressed in the inline threads above, and CI is green.

File: packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts [hunk 6]
```diff
@@ -118,3 +124,87 @@ export function analyzeA2AMentions(
...
+ * Conditions (all must hold):
+ *  1. @pattern appears mid-line (not at line start)
+ *  2. Action keyword immediately adjacent to @mention (proximity-based, not whole-line)
+ *  3. Not inside a fenced code block or blockquote
+ *  4. Target cat was not already routed via line-start mention
+ *  5. Not a self-mention
+ */
+
+/** Action patterns that appear immediately BEFORE @mention (e.g. "Ready for @xxx"). */
+const BEFORE_HANDOFF_RE = /(?:ready\s+for|交接给?|转给|请|帮)\s*$/i;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts]: Fixed in `8b7f65f`. Added negative lookbehind `(?<![邀申敬])请` to `BEFORE_HANDOFF_RE`:

- `邀请 @codex` → 邀请 = invite → no trigger ✅
- `申请 @codex` → 申请 = apply → no trigger ✅
- `敬请 @codex` → 敬请 = compound → no trigger ✅
- `请 @codex review` → standalone 请 → still triggers ✅

3 new tests added, 58/58 pass.

*[宪宪/Opus-46🐾]*

**Comment by mindfn** [comment]
Re: Codex P2 — already-seen cat blocking fresh cats on same line

Fixed in `6d5b0f2`.

**Root cause**: `lineMatched = true` was set unconditionally after proximity match, even when the cat was already in `seen`. Deduped cats "claimed" the line and blocked fresh cats.

**Fix**: Moved `lineMatched = true` inside the `!seen` branch — only a newly-recorded finding claims the line.

**Repro test**: `"Ready for @gemini review\nReady for @gemini and @codex review"` — @gemini (longer pattern, scanned first) was already seen from line 1; on line 2 it blocked @codex. Now both detected correctly.

59/59 pass.

*[宪宪/Opus-46🐾]*

File: packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts [hunk 7]
```diff
@@ -118,3 +124,90 @@ export function analyzeA2AMentions(
...
+        // Already routed via line-start: skip this entry but keep scanning other cats on same line.
+        if (routedSet.has(entry.catId)) break;
+        const before = normalized.slice(0, idx);
+        const after = normalized.slice(idx + entry.pattern.length);
+        if (!BEFORE_HANDOFF_RE.test(before) && !AFTER_HANDOFF_RE.test(after)) continue;
+        if (!seen.has(entry.catId)) {
+          seen.add(entry.catId);
+          found.push({ catId: entry.catId, lineText: rawLine.trim() });
+        }
+        lineMatched = true;
```
[the remark this answers is not in the document]
Comment (mindfn) [reply packages/api/src/domains/cats/services/agents/routing/a2a-mentions.ts]: Fixed in `6d5b0f2`. Moved `lineMatched = true` inside the `!seen` branch so deduped cats don't claim the line. Added regression test: `"Ready for @gemini review\nReady for @gemini and @codex review"` — gemini (longer pattern, scanned first) was already seen from line 1; now codex on line 2 is correctly detected. 59/59 pass.

*[宪宪/Opus-46🐾]*
